getblocklab/block-lab
# In the migration, install and activate Genesis Custom Blocks

#### Changes
* As John suggested, changes the order of the GCB installation and block migration
* Installs and activates GCB via a REST API endpoint
* The endpoint is mainly taken from the Gutenberg [plugin installation route](https://github.com/WordPress/gutenberg/blob/8d64aa3092d5d9e841895bf2d495565c9a770238/lib/class-wp-rest-plugins-controller.php#L417)
* Ideally the JS file would just use that endpoint, but it's only in Gutenberg, not in Core

#### Testing instructions (updated 23 July)

1. Checkout this PR's branch, and also checkout https://github.com/studiopress/genesis-custom-blocks/pull/13, but don't activate GCB it yet
2. Go to `/wp-admin` > Block Lab > Migrate, and go through the process
3. Expected: 
![expected-diff](https://user-images.githubusercontent.com/4063887/88240490-e765a300-cc4c-11ea-83ff-bb40ffa6c0f4.gif)

4. Change the slug to `amp`, as that's available on wp.org to download:

```diff
diff --git a/php/admin/migration/class-api.php b/php/admin/migration/class-api.php
index 9d4b3ae..7e36d98 100644
--- a/php/admin/migration/class-api.php
+++ b/php/admin/migration/class-api.php
@@ -45,7 +45,7 @@ class Api extends Component_Abstract {
         * Api constructor.
         */
        public function __construct() {
-               $this->new_plugin_slug = 'genesis-custom-blocks';
+               $this->new_plugin_slug = 'amp';
                $this->new_plugin_file = "{$this->new_plugin_slug}/{$this->new_plugin_slug}.php";
        }
```
GCB is not on wp.org of course, so we can't really test downloading and installing it.

5. Go through the migration flow just like before
6. Expected: The plugin should have been installed and activated:

<img width="726" alt="plugin-activated" src="https://user-images.githubusercontent.com/4063887/88241075-ab334200-cc4e-11ea-9b33-df3b6c1ff08b.png">

7. To reverse the migration, you might need to do this:
```sh
wp shell
( new Block_Lab\Admin\Migration\Post_Content( 'amp', 'block-lab' ) )->migrate_all();
```

8. Here are more steps to [reverse the migration](https://github.com/getblocklab/block-lab/pull/629#issuecomment-655846402)

9. If you have time, testing [these steps](https://github.com/getblocklab/block-lab/pull/629#issue-439492116) again would help, verifying the 'Expected' steps 8 and 9

Thanks for your patience 😄
